### ISSUE TITLE
Code Review Requested for EVERYTHING-CHANGES

### ISSUE THREAD

**Issue by lillian-lemmer** [hypatia-engine/hypatia#70]
**Notice/edit:** I'll just keep updating this initial issue as debug information changes and updates unless there's a MAJOR announcement.

**Last updated**: 2015-09-14T16:44-0600

**[@ejmr's ongoing code review](https://gist.github.com/ejmr/3cc404387af3bff2c7db)**

## Summary

I am requesting an ongoing code review of the branch `EVERYTHING-CHANGES`. There are a series of related, tricky bugs.

I made an `EVERYTHING-CHANGES` branch of Hypatia. The idea is to replace `pyganim` as a dependency, with an animation system that's far quicker. I have succeeded in that: `SpriteAnimation` works in a separate `pygame`/`SpriteAnimation` test, far quicker than `pyganim`. This changed a lot of code which deals with animations, rendering stuff, etc. Basically, again, `EVERYTHING-CHANGES`. Here are some other changes introduced:

  * Updated docstrings
  * `physics.AbsolutePosition`
  * `physics.Velocity`
  * Changes to constants
  * ???

## Here are the issues

  1. Anchored hat walkabout is not rendering (seemingly at all?)
  2. Impassability/collision seems very broken, I'd estimate from implementing the new absoluteposition system
  3. Animations don't seem to be... at all animated. However, when I take this system outside of Hypatia, it seems to work just fine! Upon request I can show a functional demo of the animation system in a quick pygame demo.

## Other notes

**Please debug in Python 2.**

When printing the active direction and action of the player, the player alternates between walk and stand; I tested this by placing a couple of print statements in `player.Player.move()`. This could explain why "animations are not working." I have verified that animations work outside the scope of Hypatia, so it's not the animation code.

The bug where the camera/viewport jitters as you make it move is caused by some floor division discrepancy between Python 2 and Python 3. Right now I'm focusing on the other issues and am focusing on Python 2.

Areas I'm focusing on: collision detection, `player.move()`, `physics.AbsolutePosition`, `game.Game`, `animations.Walkabout`.

To clear confusion `self.rect` is actually used for screen position, but never for absolute position.

I think there may be a series of bugs related to using the aformentioned "screen position" `self.rect` for various collision detection, etc.

**Comment by ejmr** [comment]
I will work on a review in my free time over the next few days and send you some information no later than Monday.

**Comment by lillian-lemmer** [comment]
Thank you a ton, @ejmr, you have no idea how much I appreciate that. x.x It was driving me crazy and I haven't had any time to do any real debugging so it's been this source of frustration for me and I felt like it's been holding development back.

**Comment by ejmr** [comment]
Sorry, but unexpectedly being bed-ridden for the past few days has held me back on this task.  I'm just now starting my review of the whole code base, so it will probably be a few days before I finish it and put all my notes in order to send your way.

**Comment by lillian-lemmer** [comment]
@ejmr don't worry about it, sweetie. :) I really appreciate the effort. I'm sorry you haven't been doing well. :/

**Comment by ejmr** [comment]
Thanks :smile: 

Reading/writing code is my favorite thing to do to help feel better.  I also have to admit there is something funny to me about running `git checkout EVERYTHING-CHANGES` from my terminal, heh.

**Comment by lillian-lemmer** [comment]
<3 Aw, same!

LOL, yeah I was in a *funny* mood when I made that branch name. :smile:

**Comment by ejmr** [comment]
[My notes as I go through the code.](https://gist.github.com/ejmr/3cc404387af3bff2c7db).

**Comment by lillian-lemmer** [comment]
If I print the current action and direction while moving (print player.active_action and player.active_direction) it seems the player alternates between walking, standing every-other frame.

screen shaking bug does not occur in python2

I will try to keep the comments of findings in here rather than in your notes :p 

Here are some notes from therein:

i think "Line 238: A comment mentions that self.rect is never used." is just outdated--I believe it's because self.rect is actually used for screen position, but never for absolute position.

i think there may be a series of bugs related to using the aformentioned "screen position" self.rect for various collision detection, etc.

**Comment by lillian-lemmer** [comment]
I found out why animations aren't working. This line:

https://github.com/hypatia-engine/hypatia/blob/EVERYTHING-CHANGES/hypatia/animations.py#L316

Was indeed returning the timedelta between two clock updates, but it was also updating the clock itself! So when replaced with `clock.get_time()` instead of `clock.tick()` it works. :) There's still some impassibility issues, but I've already started a new branch... will keep y'all updated.
